### PR TITLE
Refactor most of `kubectl drain` as a library

### PR DESCRIPTION
**Pull request by errordeveloper**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Currently draining a node is only possible with `kubectl drain`. Other projects, e.g. [cluster autoscaler](https://github.com/kubernetes/autoscaler/blob/dadf2728eb17dc2da821a09c166b036b145163d3/cluster-autoscaler/utils/drain/drain.go) and [openshift](https://github.com/openshift/kubernetes-drain/blob/c2e51be1758efa30d71a4d30dc4e2db86b70a4df/drain.go) have reinvented some of this code. This PR moves what can be moved easily from `pkg/kubectl/cmd/drain` into `pkg/util/node/drain`.

This is intended as a simple code move, there are no changes to in behaviour.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/util/node/drain/drain.go
Comment (errordeveloper) [inline pkg/util/node/drain/drain.go]: This is a little odd side-effect, I'll see how I can get rid of it easily.

File: pkg/kubectl/cmd/drain/drain.go
Comment (errordeveloper) [inline pkg/kubectl/cmd/drain/drain.go]: This function wants to write output a lot, I've not moved it cause I believe ideally libraries shouldn't write output. I've not figured out how alternative version of this would look like, and decide to not do that for the purpose of first PR.

**Comment by wking** [comment]
It's going to take me a bit to work through the diff of this vs. [openshift/kubernetes-drain][1], but just comparing the APIs, it looks like you've decided to not include (un)cordon functions?  Are you expecting to do that in follow-up work?  It seems like they'd be important for folks trying to use a drain library.

[1]: https://godoc.org/github.com/openshift/kubernetes-drain

**Comment by errordeveloper** [comment]
> APIs, it looks like you've decided to not include (un)cordon functions? Are you expecting to do that in follow-up work? 

Yes, indeed I am intending to do that. I also need those myself, of course.

**Comment by errordeveloper** [comment]
> It's going to take me a bit to work through the diff of this vs. openshift/kubernetes-drain

Have you tried to compare it to what's in master here in k/k now?

**Comment by wking** [comment]
> > It's going to take me a bit to work through the diff of this vs. openshift/kubernetes-drain
>
> Have you tried to compare it to what's in master here in k/k now?

From openshift/kubernetes-drain@2b21795ef, you can see that I split off from the then-current d43e1b3.  Changes since then:

```console
$ git describe --always origin/master
v1.14.0-alpha.0-1656-gdc6f3d6
$ git log --oneline --decorate d43e1b3..dc6f3d6 -- pkg/kubectl/cmd/drain.go
33adf36 Move each kubectl command to a separate directory
97b2992 Update gofmt for go1.11
fecb5ed Merge pull request #66266 from wking/kubectl-drain-drop-backOff
686f29f (origin/pr/69438) Merge pull request #66301 from wking/kubectl-drain-drop-typer
f008365 Merge pull request #68806 from seans3/legacy-scheme-update
ab993e3 (origin/pr/66266) kubectl: Drop backOff from DrainOptions
7a2a987 Move legacyscheme (internal version) to kubectl scheme (external version)
452615c (origin/pr/68767) Fix drain for evicting terminal DS pods and pods with local storage
967280b (origin/pr/68069) Add --server-dry-run flag to `kubectl apply`
5b55e1f (origin/pr/67658) Create cli-runtime staging repository
0d81e83 (origin/pr/66301) kubectl: Drop typer from DrainOptions
```

#66266 and #66301 ported some of my downstream changes back upstream here.  I'll take a close look at #68767, #68806, and the others when I get a moment.

**Comment by errordeveloper** [comment]
I am really not sure why this error appears in tests:
```
            error: unable to cordon node "node": no kind "Node" is registered for the internal version of group "" in scheme "pkg/kubectl/scheme/scheme.go:28"
```

I assume it breaks tests as output has changed, but I worry that the above error was there before this change.

**Comment by neolit123** [comment]
i think this can be fixed by just calling ./hack/update-bazel.sh

```
W0122 13:20:21.363] GoCompile: missing strict dependencies:
W0122 13:20:21.363] 	/bazel-scratch/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/sandbox/linux-sandbox/2345/execroot/__main__/pkg/kubectl/cmd/drain/drain.go: import of "k8s.io/client-go/rest"
```

File: pkg/util/node/drain/filters.go
Comment (neolit123) [inline pkg/util/node/drain/filters.go]: 2019

File: pkg/util/node/drain/filters.go
Comment (neolit123) [inline pkg/util/node/drain/filters.go]: for consistency, i think it would make sense to still make this function as a member of `Options` even if it does not use the object fields.

File: pkg/util/node/drain/filters.go
Comment (neolit123) [inline pkg/util/node/drain/filters.go]: non-blocking, nit as this is a copy-paste from the old code, but this should have been a comparison against a GVK -> Kind instead of a comparison to a string.

File: pkg/util/node/drain/drain.go
Comment (neolit123) [inline pkg/util/node/drain/drain.go]: 2019

File: pkg/util/node/drain/drain.go
Comment (neolit123) [inline pkg/util/node/drain/drain.go]: would it be safe to remove this field at this point?

File: pkg/kubectl/cmd/drain/drain.go
Comment (neolit123) [reply pkg/kubectl/cmd/drain/drain.go]: backends shouldn't be necessary on mute; they should allow piping to a custom stream.

i see this is already possible using `o.Out/ErrOut`, yet if this is always printing to the standard streams and if there is no easy way to customize it, it needs a minor refactor.
but given `o.Out/ErrOut` exists i'd consider moving this function too.

**Review by neolit123 — COMMENTED** [review]
thanks for the updates @errordeveloper 
added some more comments.

File: pkg/util/node/drain/drain.go
Comment (errordeveloper) [reply pkg/util/node/drain/drain.go]: Yes, indeed.

File: pkg/kubectl/cmd/drain/drain.go
Comment (errordeveloper) [reply pkg/kubectl/cmd/drain/drain.go]: Yes, but the streams are `genericclioptions.IOStreams`, and those are not convenient to use outside of Kubernetes project itself, as an additional wrapping is needed to attach those to a logger or something else.

Comment (neolit123) [reply pkg/kubectl/cmd/drain/drain.go]: i see, so that's a problem, yes.
consumers should have an easy way to manage the streams.

**Comment by neolit123** [comment]
/skip
we were instructed to add skip to PRs that are blocking on `pull-kubernetes-e2e-kops-aws`
it was removed from presubmits, so old PRs might be stuct still trying to test it.

**Comment by errordeveloper** [comment]
I've fix two bugs and the unit tests are passing locally now.

File: pkg/kubectl/cmd/drain/drain.go
Comment (errordeveloper) [reply pkg/kubectl/cmd/drain/drain.go]: Yes, so I'm think of addressing this in a follow-up PR, what do you think?

File: pkg/util/node/drain/filters.go
Comment (errordeveloper) [reply pkg/util/node/drain/filters.go]: 👍

**Comment by errordeveloper** [comment]
This is getting into a better shape now, @neolit123 it'd be great if you could take another look! There is one issue right now that comes up in manual tests - duplicate output lines.

**Comment by errordeveloper** [comment]
Unit tests seem to panic when ran via bazel... hm!

**Comment by errordeveloper** [comment]
I've fixed the race condition that bazel tests detected, which was to do with deletions lists containing non-unique items. I've also added an assertion in the tests that should prevent same from happening again.

**Comment by errordeveloper** [comment]
If the tests pass in CI, I'm happy with this as it stands and would like to hear what reviewers/approvers think of this.

File: pkg/kubectl/cmd/drain/drain.go
Comment (neolit123) [reply pkg/kubectl/cmd/drain/drain.go]: sounds good!

**Review by neolit123 — APPROVED** [review]
LGTM on my side @errordeveloper 
thank you!

**Comment by errordeveloper** [comment]
cc @kubernetes/sig-cli-maintainers

**Review by soltysh — CHANGES_REQUESTED** [review]
/hold
This PR is destroying the progress sig-cli has been going to split kubectl out of main repo. For that we CANNOT rely on any parts from the main repo. You can copy the functionality but you cannot make kubectl rely on any package from within k/k repo.

**Comment by errordeveloper** [comment]
Hi @soltysh! I certainly had no intention of destroying any kind of progress. Let's find a better approach to this, I'm happy to connect here or on Slack. And if you insist, in a SIG meeting otherwise.

**Comment by errordeveloper** [comment]
/assign @soltysh

As we discussed on Slack, I've moved the new package under `pkg/kubectl/drain`. If there is anything else that I should do, please let me know. The PR still only does a partial refactoring, and it seems plausible to finish it all up in a follow-up PR. With regards to tests, there aren't any tests yet for the new package, it's only covered by existing unit tests in `pkg/kubectl/cmd/drain` and whatever e2e tests there that cover `kubectl drain` functionality as such.

**Review by soltysh — APPROVED** [review]
/approve

**Comment by soltysh** [comment]
Please squash your changes and I'll push that forward.

**Comment by awh** [comment]
Nice work @errordeveloper - once I can use this in `kured` I no longer have to include `kubectl` in the image, which will reduce its size significantly.

**Review by soltysh — APPROVED** [review]
/lgtm

File: pkg/kubectl/cmd/drain/drain.go
Comment (tedyu) [inline pkg/kubectl/cmd/drain/drain.go]: Should newErrs be aggregated into err ?

Comment (errordeveloper) [reply pkg/kubectl/cmd/drain/drain.go]: You mean from the library call? That can be discussed. There are still quite a few improvements to be made.

Comment (tedyu) [reply pkg/kubectl/cmd/drain/drain.go]: What do you think of the following ?
```
diff --git a/pkg/kubectl/cmd/drain/drain.go b/pkg/kubectl/cmd/drain/drain.go
index 8e98e84ee6..76beaef2cf 100644
--- a/pkg/kubectl/cmd/drain/drain.go
+++ b/pkg/kubectl/cmd/drain/drain.go
@@ -322,6 +322,8 @@ func (o *DrainCmdOptions) deleteOrEvictPodsSimple(nodeInfo *resource.Info) error
                }
                if newErrs != nil {
                        fmt.Fprintf(o.ErrOut, "following errors also occurred:\n%s", utilerrors.NewAggregate(newErrs))
+                       newErrs = append(newErrs, err)
+                       err = utilerrors.NewAggregate(newErrs)
                }
                return err
        }
```

Comment (errordeveloper) [reply pkg/kubectl/cmd/drain/drain.go]: It could be good, but I am not entirely sure, as the intention here was to avoid drastic changes of kubectl output, or general behaviour. This code originally returned the first error and printed new errors, so I kept it the same way. Not say it has to be that way, please open another PR if you strongly believe this must change. Small PR would be easy to discuss.

**Pull request merged**
